design/worker-pods.md

```diff
@@ -0,0 +1,161 @@
+    - /velero
+    - backup
+    - run
+    env:
+    - name: VELERO_SCRATCH_DIR
+      value: /scratch
+    - name: AWS_SHARED_CREDENTIALS_FILE
+      value: /credentials/cloud
+    image: <VELERO_SERVER_IMAGE>
+    imagePullPolicy: IfNotPresent
+    name: velero
+    volumeMounts:
+    - mountPath: /plugins
+      name: plugins
+    - mountPath: /scratch
+      name: scratch
+    - mountPath: /credentials
+      name: cloud-credentials
+    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
+      name: velero-token-plwsw
+      readOnly: true
+  serviceAccount: velero
+  volumes:
+  - emptyDir: {}
+    name: plugins
+  - emptyDir: {}
+    name: scratch
+  - name: cloud-credentials
+    secret:
+      defaultMode: 420
+      secretName: cloud-credentials
+  - name: velero-token-plwsw
+    secret:
+      defaultMode: 420
+      secretName: velero-token-plwsw
+```
+
+### Updates to `velero backup logs`
+
+The `velero backup logs` command will be modified so that if a backup is in progress, the logs are gotten from the worker pod's stdout (essentially proxying to `kubectl -n velero logs POD_NAME`). A  `-f/--follow` flag will be added to allow streaming of logs for an in-progress backup.
+
+Once a backup is complete, the logs will continue to be uploaded to object storage, and `velero backup logs` will still fetch them from there.
+
+### Changes to backup deletion controller
+
+The Velero server keeps a record of current in-progress backups and disallows them from being deleted. This will need to be updated to account for backups running in worker pods. The most sensible approach is for the backup deletion controller to first check the backup's status via the API, and to allow deletion to proceed if it's not new/in-progress. If the backup is new or in-progress, the controller can check to see if a worker pod is running for the backup. If so, the deletion is disallowed because the backup is actively being processed. If there is no worker pod for the backup, the backup can be deleted as it's not actively being processed.
+
+### Open items
+
+- It probably makes sense to use Kubernetes Jobs to control worker pods, rather than directly creating "bare pods". The Job will ensure that a worker pod successfully runs to completion.
+- Currently, restic repository lock management is handled by an in-process lock manager in the Velero server. In order for backups/restores to safely run concurrently, the design for restic lock management needs to change. There is an [open issue](https://github.com/heptio/velero/issues/1540) for this which is currently scoped for the v1.1 release.
+- There are several prometheus metrics that are emitted as part of the backup process. Since backups will no longer be running in the Velero server, we need to find a way to expose those values. One option is to store any value that would feed a metric as a field on the backup's `status`, and to have the Velero server scrape values from there for completed backups.
```

Comment:
I think we should do some more research on how distributed metrics are tracked with prometheus. It may be wishful thinking, but we can't be the only ones to want to do this.

I also think the public API on our `metrics` package can be sufficient for masking most of this from the backup code, though the logic inside may change to route the metric correctly.

Reply:
potentially relevant: https://prometheus.io/docs/practices/pushing/